pkg/storage/tsdb/multilevel_chunk_cache.go

```diff
@@ -0,0 +1,149 @@
+package tsdb
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/cacheutil"
+)
+
+type multiLevelChunkCache struct {
+	name   string
+	caches []cache.Cache
+
+	backfillProcessor    *cacheutil.AsyncOperationProcessor
+	fetchLatency         *prometheus.HistogramVec
+	backFillLatency      *prometheus.HistogramVec
+	storeDroppedItems    prometheus.Counter
+	backfillDroppedItems prometheus.Counter
+	maxBackfillItems     int
+	backfillTTL          time.Duration
+}
+
+type MultiLevelChunkCacheConfig struct {
+	MaxAsyncConcurrency int `yaml:"max_async_concurrency"`
+	MaxAsyncBufferSize  int `yaml:"max_async_buffer_size"`
+	MaxBackfillItems    int `yaml:"max_backfill_items"`
+
+	BackFillTTL time.Duration `yaml:"-"`
+}
+
+func (cfg *MultiLevelChunkCacheConfig) Validate() error {
+	if cfg.MaxAsyncBufferSize <= 0 {
+		return errInvalidMaxAsyncBufferSize
+	}
+	if cfg.MaxAsyncConcurrency <= 0 {
+		return errInvalidMaxAsyncConcurrency
+	}
+	if cfg.MaxBackfillItems <= 0 {
+		return errInvalidMaxBackfillItems
+	}
+	return nil
+}
+
+func (cfg *MultiLevelChunkCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
+	f.IntVar(&cfg.MaxAsyncConcurrency, prefix+"max-async-concurrency", 50, "The maximum number of concurrent asynchronous operations can occur when backfilling cache items.")
```

Comment:
Let's reduce the backfill concurrency? 50 is a bit high. We should adjust the config for index cache as well in a separate PR.
2 ~ 4 might be a reasonable number. I have seen 50 goroutines doing backfilling taking a lot of CPUs.

Reply:
I pick the median. :)